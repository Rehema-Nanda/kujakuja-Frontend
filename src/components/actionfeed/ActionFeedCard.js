import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Row, Col, Card, CardImg } from "reactstrap";
import moment from "moment";
import PropTypes from "prop-types";
import convert from "htmr";
import i18n from "i18next";

import AppConfig from "../../AppConfig";
import { getIdeafeedParamsFromUrl } from "../../helpers";

const googleTranslateKey = AppConfig.GOOGLE_TRANSLATE_API_KEY;
const googleTranslate = require("google-translate")(googleTranslateKey);
const googleTranslateBadge = require("../../img/google_translate_greyscale.svg");

class ActionFeedCard extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isDetailsAccordionOpen: false,
            translatedDescription: "",
            translatedTitle: "",
            originalActionLang: "",
            startDate: "",
            endDate: ""
        };
        // this.formatStartEndIdeafeedDates();
    }

    componentDidMount() {
        const { action } = this.props;
        if (i18n.language !== action.action_feed_language) {
            this.translateAction();
        }
    }

    translateAction = () => {
        const { action} = this.props;
        const currentLang = i18n.language;

        googleTranslate.translate([action.title, action.description], action.action_feed_language, currentLang, (err, translate) => {
            if (err) {
                return;
            }

            this.setState({ 
                translatedTitle: translate[0].translatedText,
                translatedDescription: translate[1].translatedText
            });
        });
    }

    generateTranslationMarkUp = () => {
        const { action } = this.props;

        return (
            `${i18n.language}-x-mtfrom-${action.action_feed_language}`
        );
    };

    toggleDetailsAccordion = () => {
        const { isDetailsAccordionOpen } = this.state;
        this.setState({ isDetailsAccordionOpen: !isDetailsAccordionOpen });
    };

    formatStartEndIdeafeedDates = () => {
        const period = {};
        const { action } = this.props;
        const params = getIdeafeedParamsFromUrl(action.source);
        const startDate = moment(params.startDate);
        const endDate = moment(params.endDate);

        startDate.isValid() ? period.startDate = startDate.format("MMMM") : period.startDate = "";
        endDate.isValid() ? period.endDate = endDate.format("MMMM, YYYY") : period.endDate = "" ;
        return period;
    }

    render() {
        const { action } = this.props;
        const { translatedDescription, translatedTitle, isDetailsAccordionOpen, originalActionLang } = this.state;

        return (
            <div className="action-card">
                <div className="action-top-details">
                    <Row>
                        <Col md="4">
                            <Card className="action-img">
                                <CardImg top width="100%" src={action.image} alt="" />
                            </Card>
                        </Col>
                        <Col md="8">
                            <div className="action-card-body">
                                <div className="action-content">
                                    <p className="action-time">
                                        {moment(action.time).format("MMMM, YYYY")}
                                        {" "}
                                        <span>&#124;</span>
                                    </p>
                                    <p className="action-area">
                                        {" "}
                                        {`${action.location}, ${action.country}`}
                                        {" "}
                                        <span>&#124;</span>
                                    </p>
                                    <p className="action-sector">
                                        {" "}
                                        {`${action.service_type}`}
                                    </p>   
                                </div>

                                <div className="action-details" lang={this.generateTranslationMarkUp()}>
                                    <h3 className="action-title">{translatedTitle || action.title}</h3>
                                    <br />
                                    <p className="action-author">
                                        {i18n.t("actionFeed.by")}
                                        {" "}
                                        {action.implementor}
                                    </p>
                                    <p className="action-desc">{convert(translatedDescription || action.description)}</p>
                                </div>
                            </div>
                        </Col>
                    </Row>
                </div>
                {i18n.language !== action.action_feed_language
                && (
                    <>
                        <div className="translation-container">
                            <div className="text">
                                <span
                                    className="show-original"
                                    role="presentation"
                                    onClick={() => {
                                        this.toggleDetailsAccordion();
                                    }}
                                >
                                    {isDetailsAccordionOpen ? i18n.t("actionFeed.hideOriginal") : i18n.t("actionFeed.showOriginal")}
                                    {" "}
                                    {originalActionLang}
                                    {" "}
                                    {i18n.t("actionFeed.action")}
                                </span>
                            </div>
                            <div className="google-badge">
                                <a href="https://translate.google.com/" rel="noopener noreferrer" target="_blank">
                                    <img src={googleTranslateBadge} alt="" />
                                </a>
                            </div>
                        </div>
                        {isDetailsAccordionOpen
                        && (
                            <div className="ideafeed-item-details">
                                <div className="ideafeed-item-details-idea">
                                    <div className="header">
                                        <span className="header-text">Action as captured</span>
                                    </div>
                                    <p className="show-original-idea title">
                                        {`${action.title}:`}
                                    </p>
                                    <p className="show-original-idea description">
                                        {convert(action.description)}
                                    </p>
                                </div>
                            </div>
                        )}
                    </>
                )}

                <div className="action-footer">
                    <Row>
                        <Col xs="8">
                            <div className="footer-desc">
                                <p>
                                    {i18n.t("actionFeed.actionTaken")}
                                    {" "}
                                    {`"${action.tag}"`}
                                    {" "}
                                    {i18n.t("actionFeed.during")}
                                    {" "}
                                    {`${this.formatStartEndIdeafeedDates().startDate} - ${this.formatStartEndIdeafeedDates().endDate}`}
                                </p>
                            </div>

                        </Col>
                        <Col xs="4">
                            <div className="action-links">
                                <a className="action-icon" href={action.source} target="_blank" rel="noopener noreferrer">
                                    <span className="footer-text">
                                        {i18n.t("actionFeed.view")}
                                        {" "}
                                        {action.ideas_count}
                                        {i18n.t("actionFeed.ideas")}
                                    </span>
                                    <FontAwesomeIcon icon="arrow-right" />
                                </a>
                            </div>
                        </Col>
                    </Row>
                </div>
            </div>
        );
    }
};

ActionFeedCard.propTypes = {
    action: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default ActionFeedCard;
