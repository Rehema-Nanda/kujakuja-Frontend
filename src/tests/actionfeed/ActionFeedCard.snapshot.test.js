import React from "react";
import { mount } from "enzyme";
import ActionFeedCard from "../../components/actionfeed/ActionFeedCard";

describe("ActionFeedCard component", () => {
    const props = {
        action: {
            location: "Mahama",
            country: "Rwanda",
            description: "there were no materials for a sustainable solution. Therefore, a Technical Team has been sent to build those stairs to help customers be safe from the accidents the place was causing them and now they are comfortably fetching water standing on that side.there were no materials for a sustainable solution. Therefore, a Technical Team has been sent to build those stairs to help customers be safe from the accidents the place was",
            image: "https://firebasestorage.googleapis.com/v0/b/kuja-kujans-survey-ecdb7.appspot.com/o/img3.jpg?alt=media&token=2627c138-258f-42d4-8ccf-254fb932b778",
            implementor: "John Doe",
            numbers: "235",
            service_type: "Nutrition Service",
            source: "https://dev.alight.kujakuja.com/en/ideafeed?start=2020-10-01&end=2020-10-31&settlements=16&keyword=water&limit=50&page=1",
            tag: "#porridge",
            time: "2020-10-11",
            title: "The ALIGHT Health Service Team has implemented an action"
        },
    };
    const wrapper = mount(<ActionFeedCard {...props} />);

    it("renders correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });
});
