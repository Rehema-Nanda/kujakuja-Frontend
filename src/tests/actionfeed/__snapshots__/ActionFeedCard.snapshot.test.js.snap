// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionFeedCard component renders correctly 1`] = `
"<ActionFeedCard action={{...}}>
  <div className=\\"action-card\\">
    <div className=\\"action-top-details\\">
      <Row tag=\\"div\\">
        <div className=\\"row\\">
          <Col md=\\"4\\" tag=\\"div\\" widths={{...}}>
            <div className=\\"col-md-4\\">
              <Card className=\\"action-img\\" tag=\\"div\\">
                <div className=\\"action-img card\\">
                  <CardImg top={true} width=\\"100%\\" src=\\"https://firebasestorage.googleapis.com/v0/b/kuja-kujans-survey-ecdb7.appspot.com/o/img3.jpg?alt=media&token=2627c138-258f-42d4-8ccf-254fb932b778\\" alt=\\"\\" tag=\\"img\\">
                    <img width=\\"100%\\" src=\\"https://firebasestorage.googleapis.com/v0/b/kuja-kujans-survey-ecdb7.appspot.com/o/img3.jpg?alt=media&token=2627c138-258f-42d4-8ccf-254fb932b778\\" alt=\\"\\" className=\\"card-img-top\\" />
                  </CardImg>
                </div>
              </Card>
            </div>
          </Col>
          <Col md=\\"8\\" tag=\\"div\\" widths={{...}}>
            <div className=\\"col-md-8\\">
              <div className=\\"action-card-body\\">
                <div className=\\"action-content\\">
                  <p className=\\"action-time\\">
                    October, 2020
                     
                    <span>
                      |
                    </span>
                  </p>
                  <p className=\\"action-area\\">
                     
                    Mahama, Rwanda
                     
                    <span>
                      |
                    </span>
                  </p>
                  <p className=\\"action-sector\\">
                     
                    Nutrition Service
                  </p>
                </div>
                <div className=\\"action-details\\" lang=\\"undefined-x-mtfrom-undefined\\">
                  <h3 className=\\"action-title\\">
                    The ALIGHT Health Service Team has implemented an action
                  </h3>
                  <br />
                  <p className=\\"action-author\\">
                     
                    John Doe
                  </p>
                  <p className=\\"action-desc\\">
                    there were no materials for a sustainable solution. Therefore, a Technical Team has been sent to build those stairs to help customers be safe from the accidents the place was causing them and now they are comfortably fetching water standing on that side.there were no materials for a sustainable solution. Therefore, a Technical Team has been sent to build those stairs to help customers be safe from the accidents the place was
                  </p>
                </div>
              </div>
            </div>
          </Col>
        </div>
      </Row>
    </div>
    <div className=\\"action-footer\\">
      <Row tag=\\"div\\">
        <div className=\\"row\\">
          <Col xs=\\"8\\" tag=\\"div\\" widths={{...}}>
            <div className=\\"col-8\\">
              <div className=\\"footer-desc\\">
                <p>
                   
                  &quot;#porridge&quot;
                   
                   
                  October - October, 2020
                </p>
              </div>
            </div>
          </Col>
          <Col xs=\\"4\\" tag=\\"div\\" widths={{...}}>
            <div className=\\"col-4\\">
              <div className=\\"action-links\\">
                <a className=\\"action-icon\\" href=\\"https://dev.alight.kujakuja.com/en/ideafeed?start=2020-10-01&end=2020-10-31&settlements=16&keyword=water&limit=50&page=1\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
                  <span className=\\"footer-text\\">
                     
                  </span>
                  <FontAwesomeIcon icon=\\"arrow-right\\">
                    <i className=\\"fa\\" />
                  </FontAwesomeIcon>
                </a>
              </div>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </div>
</ActionFeedCard>"
`;
